### PR TITLE
API Uses Stripe to manage subscription

### PR DESCRIPTION
**Pull request by Fraseer**
PT Link: https://www.pivotaltracker.com/story/show/179900893

## User Story
As a news paper api
In order to get paid by subscribers
I would like process payments using card

## Changes Proposed
- Add  stripe-rails & stripe-ruby-mock gems
- Generated subscription controller
- Add subscription route
- Add `publishable_key` & `secret_key` to config

## Screenshots
image

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -12,6 +12,8 @@ gem 'simplecov', require: false, group: :test
 gem 'devise_token_auth'
 gem 'pundit'
 gem 'active_model_serializers', '~> 0.10.12'
+gem 'stripe-rails', '~> 2.3', '>= 2.3.2'
+gem 'stripe-ruby-mock', '~> 3.1.0.rc2', require: 'stripe_mock'
```
Comment (tochman) [inline Gemfile]: This needs to be placed in `:development, :test` group

File: app/controllers/api/subscriptions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
+class Api::SubscriptionsController < ApplicationController
+  before_action :authenticate_user!, only: :create
+  rescue_from Stripe::CardError, with: :rescue
+  rescue_from Stripe::InvalidRequestError, with: :rescue
+
+  def create
+    customer = Stripe::Customer.list(email: current_user.email).data.first
+    customer ||= Stripe::Customer.create(email: current_user.email, source: params[:stripeToken]) unless customer
```
Comment (tochman) [inline app/controllers/api/subscriptions_controller.rb]: `||=` and `unless customer` feels redundant. Please make sure you know what each line of code is doing.

File: app/controllers/api/subscriptions_controller.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    Stripe::Charge.create(customer: customer.id, currency: 'sek', amount: 50_000)
```
Comment (tochman) [inline app/controllers/api/subscriptions_controller.rb]: The question is if the amount will be passed in as params to this action, or is it always going to be 500SEK?

File: app/controllers/api/subscriptions_controller.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+
+    render json: { message: 'You have successfully subscribed' }, status: 201
+    current_user.subscriber!
```
Comment (tochman) [inline app/controllers/api/subscriptions_controller.rb:10]: What if this fails?

File: spec/requests/api/subscription_create_spec.rb [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+      StripeMock.prepare_card_error(:processing_error)
+      post '/api/subscriptions', params: { stripeToken: valid_card_token }, headers: credentials
+    end
+
+    it { is_expected.to have_http_status 402 }
+
+    it 'is expected to return an error message' do
+      expect(response_json['errors']).to eq 'An error occurred while processing the card'
+    end
+  end
```
Comment (tochman) [inline spec/requests/api/subscription_create_spec.rb:77]: I would like to see a test for what happens if the user is not authenticated.

File: Gemfile (hunk 1 above)
Comment (Fraseer) [reply Gemfile]: Moved

File: spec/requests/api/subscription_create_spec.rb (hunk 5 above)
Comment (Fraseer) [reply spec/requests/api/subscription_create_spec.rb:77]: Added test for non authenticated user

File: app/controllers/api/subscriptions_controller.rb (hunk 3 above)
Comment (Fraseer) [reply app/controllers/api/subscriptions_controller.rb]: It should not be, have updated it to take a input

File: app/controllers/api/subscriptions_controller.rb (hunk 4 above)
Comment (Fraseer) [reply app/controllers/api/subscriptions_controller.rb:10]: I think that as this is going to Stripe any failures will come in the form of a Stripe::xxxxError, and therefore changing to have a `rescue_from Stripe::StripeError`(the base error from which all other more specific Stripe errors derive) at the top of the file will catch any that arise.

File: app/controllers/api/subscriptions_controller.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+class Api::SubscriptionsController < ApplicationController
+  before_action :authenticate_user!, only: :create
+  rescue_from Stripe::StripeError, with: :rescue
+
+  def create
+    customer = Stripe::Customer.list(email: current_user.email).data.first
+    customer ||= Stripe::Customer.create(email: current_user.email, source: params[:stripeToken],
+                                         currency: params[:currency], amount: params[:amount])
```
Comment (tochman) [inline app/controllers/api/subscriptions_controller.rb]: Is `amount` an attribute of the `customer`? It would make more sense it relates to the `charge` 🤔

Comment (Fraseer) [reply app/controllers/api/subscriptions_controller.rb]: Amount is one of the predefined attrs of customer,  but it does also work on the charge line

**Review by maxarvid — APPROVED** [review]
LGTM

**Review by lukeperrin10 — APPROVED** [review]

**Pull request merged**